pkg/proctree/proctree_feed.go

```diff
@@ -0,0 +1,257 @@
+			Pid:         int(feed.ChildPid),
+			NsTid:       int(feed.ChildNsTid),
+			NsPid:       int(feed.ChildNsPid),
+			StartTimeNS: feed.ChildStartTime,
+			PPid:        int(feed.ParentPid),
+			NsPPid:      int(feed.ParentNsPid),
+			Uid:         0, // TODO: implement
+			Gid:         0, // TODO: implement
+		},
+		utils.NsSinceBootTimeToTime(feed.TimeStamp),
+	)
+	thread.SetParentHash(feed.ParentHash) // all threads have the same parent as the thread group leader
+	thread.SetLeaderHash(feed.LeaderHash) // thread group leader is a "process" (not a thread)
+	leader.AddThread(feed.ChildHash)      // add the thread to the thread group leade
+
+	return nil
+}
+
+type ExecFeed struct {
+	TimeStamp         uint64
+	TaskHash          uint32
+	CmdPath           string
+	PathName          string
+	Dev               uint32
+	Inode             uint64
+	Ctime             uint64
+	InodeMode         uint16
+	InterPathName     string
+	InterDev          uint32
+	InterInode        uint64
+	InterCtime        uint64
+	Interpreter       string
+	StdinType         uint16
+	StdinPath         string
+	InvokedFromKernel int32
+}
+
+// FeedFromExec feeds the process tree with an exec event.
+func (pt *ProcessTree) FeedFromExec(feed ExecFeed) error {
+	// DEBUG (TODO: remove this)
+	// file, _ := os.Open("/dev/null")
+	// file := os.Stdout
+	// fmt.Fprintf(file, "--\nExec event received:\n")
+	// fmt.Fprintf(file, "taskHash=%v\n", feed.TaskHash)
+	// fmt.Fprintf(file, "cmdPath=%v\n", feed.CmdPath)
+	// fmt.Fprintf(file, "pathName=%v\n", feed.PathName)
+	// fmt.Fprintf(file, "interPathName=%v\n", feed.InterPathName)
+	// fmt.Fprintf(file, "interpreter=%v\n", feed.Interpreter)
+	// fmt.Fprintf(file, "stdinPath=%v\n", feed.StdinPath)
+	// fmt.Fprintf(file, "dev=%v\n", feed.Dev)
+	// fmt.Fprintf(file, "interDev=%v\n", feed.InterDev)
+	// fmt.Fprintf(file, "inode=%v\n", feed.Inode)
+	// fmt.Fprintf(file, "ctime=%v\n", feed.Ctime)
+	// fmt.Fprintf(file, "interInode=%v\n", feed.InterInode)
+	// fmt.Fprintf(file, "interCtime=%v\n", feed.InterCtime)
+	// fmt.Fprintf(file, "inodeMode=%v\n", feed.InodeMode)
+	// fmt.Fprintf(file, "stdinType=%v\n", feed.StdinType)
+	// fmt.Fprintf(file, "invokedFromKernel=%v\n", feed.InvokedFromKernel)
+	// END OF DEBUG
+
+	process, procOk := pt.GetProcessByHash(feed.TaskHash)
+	_, threadOk := pt.GetThreadByHash(feed.TaskHash)
+
+	if !procOk && !threadOk {
+		logger.Debugw("process or thread not found (evicted ?)", "taskHash", feed.TaskHash)
+		return nil
+	}
+
+	// Running execve() from a thread is discouraged and behavior can be unexpected:
+	//
+	// 1. All threads are terminated.
+	// 2. PID remains the same.
+	// 3. New process is single threaded.
+	// 4. Inherited Attributed (open fds, proc group, ID, UID, GID, ... are retained)
+	// 5. Still, it isn't forbidden and should be handled.
+
+	if threadOk {
+		// TODO: handle execve() from a thread
+		logger.Debugw("exec event received for a thread", "taskHash", feed.TaskHash)
+		return nil
+	}
```

Comment:
Why do we care that we executed from a thread? Won't it be better to update our data structures?

Reply:
Thats the "TODO", to update our structures. I have to test to see if the exits() for all existing threads (from the thread group) will occur and which remaining task keeps alive (probably the leader) and then update the leader with the new exec'ed data (so the changelog would show it being a thread group leader, and then again a thread group leader with different FileInfo data). But this needs testing and its a "corner case".

Reply:
I'll deal with this item (TODO) in a specific new issue.

Reply:
Issue: https://github.com/aquasecurity/tracee/issues/3478